### ISSUE TITLE
LPC fails with dtype=float32

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#856]
#### Description
I just tried out the freshly merged #797 and it fails with numba errors when `y` has dtype=float32.  It works fine with float64.

Tagging this for now, will push a fix later.

**Comment by bmcfee** [comment]
The compilation error happens at [this line](https://github.com/librosa/librosa/blob/master/librosa/core/audio.py#L645), due to a type mismatch:
```
E           numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Cannot unify array(float32, 1d, C) and array(float64, 1d, C) for 'fwd_pred_error', defined at /home/bmcfee/git/librosa/librosa/core/audio.py (612)
E           
E           File "librosa/core/audio.py", line 612:
E           def __lpc(y, order):
E               <source elided>
E               # iteration of the below loop
E               fwd_pred_error = y[1:]
E               ^
E           
E           [1] During: typing of assignment at /home/bmcfee/git/librosa/librosa/core/audio.py (645)
E           
E           File "librosa/core/audio.py", line 645:
E           def __lpc(y, order):
E               <source elided>
E                   fwd_pred_error_tmp = fwd_pred_error
E                   fwd_pred_error = fwd_pred_error + reflect_coeff*bwd_pred_error
E                   ^
```

I'm guessing the multiplication of `reflect_coeff * bwd_pred_error` is upcasting from float32 to float64.  We might have to rewrite this chunk to make better use of temporary arrays.

**Comment by bmcfee** [comment]
@ajweiss could you take a look at this one when you get a chance?

**Comment by ajweiss** [comment]
it'll probably take a few days to find some time, but sure!

**Comment by ajweiss** [comment]
I added some tests for this and wasn't able to reproduce.  See PR #861.  Can you attach some additional detail on the environment this failed in?  (that is, if the CI doesn't catch it)

**Comment by bmcfee** [comment]
> Can you attach some additional detail on the environment this failed in? (that is, if the CI doesn't catch it)

Hah, yes, my bad:

```python
In [2]: librosa.show_versions()                                                               
INSTALLED VERSIONS
------------------
python: 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34) 
[GCC 7.3.0]

librosa: 0.6.3

audioread: 2.1.5
numpy: 1.16.2
scipy: 1.2.1
sklearn: 0.20.3
joblib: 0.13.2
decorator: 4.3.2
six: 1.12.0
resampy: 0.2.1
numba: 0.43.1

numpydoc: 0.8.0
sphinx: 1.8.5
sphinx_rtd_theme: 0.4.3
sphinxcontrib.versioning: 2.2.1
sphinx-gallery: None
pytest: 3.10.1
pytest-mpl: None
pytest-cov: None
matplotlib: 3.0.3
```

```python
In [3]: y, sr = librosa.load(librosa.util.example_audio_file())                               

In [5]: a = librosa.lpc(y, 5)                                                                 
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-5-563b53ac7d1d> in <module>
----> 1 a = librosa.lpc(y, 5)

~/git/librosa/librosa/core/audio.py in lpc(y, order)
    586     util.valid_audio(y, mono=True)
    587 
--> 588     return __lpc(y, order)
    589 
    590 

~/miniconda3/envs/py36/envs/py36/envs/py36/lib/python3.6/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    348                 e.patch_message(msg)
    349 
--> 350             error_rewrite(e, 'typing')
    351         except errors.UnsupportedError as e:
    352             # Something unsupported is present in the user code, add help info

~/miniconda3/envs/py36/envs/py36/envs/py36/lib/python3.6/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    315                 raise e
    316             else:
--> 317                 reraise(type(e), e, None)
    318 
    319         argtypes = []

~/miniconda3/envs/py36/envs/py36/envs/py36/lib/python3.6/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Cannot unify array(float32, 1d, C) and array(float64, 1d, C) for 'fwd_pred_error', defined at /home/bmcfee/git/librosa/librosa/core/audio.py (612)

File "git/librosa/librosa/core/audio.py", line 612:
def __lpc(y, order):
    <source elided>
    # iteration of the below loop
    fwd_pred_error = y[1:]
    ^

[1] During: typing of assignment at /home/bmcfee/git/librosa/librosa/core/audio.py (645)

File "git/librosa/librosa/core/audio.py", line 645:
def __lpc(y, order):
    <source elided>
        fwd_pred_error_tmp = fwd_pred_error
        fwd_pred_error = fwd_pred_error + reflect_coeff*bwd_pred_error
        ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

**Comment by ajweiss** [comment]
Interesting, I'm able to get it to break in conda created environments, but not in those I've  created myself.  I haven't been able to track down exactly what's different there (I was hoping it'd be something simple, like the MKL stuff), but yeah it looks like the constant scalar in `reflect_coeff` is causing the problem.  The fix is probably as easy as creating that scalar at initialization and explicitly casting it, but I'd like to get a better understanding as to why it fell through the cracks in the first place.

**Comment by bmcfee** [comment]
Thanks for digging into it.  Maybe @seibert has some insight here?  I'm not sure it warrants bringing up over in the numba issues, but it does seem like something the compiler should be able to figure out for us.

**Comment by bmcfee** [comment]
Fixed by merging #861
